### PR TITLE
Bring over starter module prs + test suite setup

### PR DESCRIPTION
**Pull request by davist11**
This brings over Greg's <a href="https://github.com/vigetlabs/craft-starter-module/pulls/gregkohn">2 PRs from the starter module</a>, sets up docker for the test suite, and writes tests.

### CODE REVIEW

File: tests/fixtures/EntriesFixture.php [hunk 1]
```diff
@@ -0,0 +1,20 @@
+<?php
+
+namespace viget\base\tests\fixtures;
+
+use craft\test\fixtures\elements\EntryFixture;
+use viget\base\tests\fixtures\SectionsFixture;
+use viget\base\tests\fixtures\EntryTypesFixture;
```
Comment (olets) [inline tests/fixtures/EntriesFixture.php]: This is necessary? Thought the `$depends` would be enough

Same question in the EntryTypesFixture file, SectionSettingsFixture file, and SectionFixture file

File: tests/functional/PartialCest.php [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+        $I->amOnPage('/partial-test');
+        $I->see('Hello World');
+    }
+
+    public function testTwigFunctionNoData(FunctionalTester $I)
+    {
+        $I->amOnPage('/partial-test');
+        $I->dontSee('This text should not be available in the partial');
+    }
+}
```
Comment (olets) [inline tests/functional/PartialCest.php:20]: Awesome

File: composer.json [hunk 3]
```diff
@@ -20,17 +20,27 @@
...
+        "codeception/lib-asserts": "^1.1.0",
+        "codeception/module-yii2": "^1.1.0",
+        "phpunit/phpunit": "^7.5.20",
+        "codeception/module-phpbrowser": "^1.0.0"
+    },
     "autoload": {
         "psr-4": {
-          "viget\\base\\": "src/"
+            "viget\\base\\": "src/",
+            "viget\\base\\tests\\fixtures\\": "tests/fixtures/"
```
Comment (olets) [inline composer.json]: I find this namespace's name unintuitive. Makes me expect the path `<viget\base's path>/tests/fixtures`

File: README.md [hunk 4]
```diff
@@ -60,4 +80,10 @@ Regardless of logged-in state or permissions, the Craft Debug Bar will always be
...
 
-When visiting an element URL on the front-end, an edit entry link to edit that element in the CP will be present in the lower left hand corner. This will always be present in the dev environment regardless of logged-in state or permissions.
\ No newline at end of file
+When visiting an element URL on the front-end, an edit entry link to edit that element in the CP will be present in the lower left hand corner. This will always be present in the dev environment regardless of logged-in state or permissions.
+
+### Components exposed to Twig `craft.viget` sub-object
+
+#### `craft.viget.util.fillInEntries()`
+
+This method queries for additional, deduped entries to a fill a number of needed entries
```
Comment (olets) [inline README.md:89]: How about documenting the parameters, and mentioning that if more than `limit` `entries` are passed in you get them all back not just `limit` of them (a good flexible behavior but worth calling out)

File: tests/unit/FillInEntriesTest.php [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    {
+        $selectedEntries = Entry::findAll([
+            'section' => 'testSection1',
+            'limit' => 4,
+        ]);
+
+        $entries = Module::$instance->util->fillInEntries($selectedEntries, $this->params, 4);
+
+        $this->assertEquals(count($entries), 4);
+    }
```
Comment (olets) [inline tests/unit/FillInEntriesTest.php:62]: Should we add a "more than enough" test?

**Review by olets — APPROVED** [review]
Very cool

Is codeception.yml about supporting the x.suite.yml files or is it something about Docker? Can see myself referring to this PR when setting up Docker and when setting up Codeception, want to make sure I know what changes are for which

Also interested in hearing how you settled on `craft.viget.util` as opposed to `viget.util`. (Fancy! I'm not opposed, just seems like a thing you must have had to _choose_)

File: tests/fixtures/EntriesFixture.php (hunk 1 above)
Comment (davist11) [reply tests/fixtures/EntriesFixture.php]: the depends might be enough, will have to test

File: README.md (hunk 4 above)
Comment (davist11) [reply README.md:89]: Need to figure out documentation overall, this is just carried over from the starter module. Likely will go a docs folder route

File: tests/unit/FillInEntriesTest.php (hunk 5 above)
Comment (davist11) [reply tests/unit/FillInEntriesTest.php:62]: Hmm not sure I follow

Comment (olets) [reply tests/unit/FillInEntriesTest.php:62]: like

```php
$selectedEntries = Entry::findAll([
    'section' => 'testSection1',
    'limit' => 5,
]);
$entries = Module::$instance->util->fillInEntries($selectedEntries, $this->params, 4);
$this->assertEquals(count($entries), 5);
```

Maybe not necessary 🤷‍♂️

Comment (gregkohn) [reply tests/unit/FillInEntriesTest.php:62]: re: your other comment and this one, is it better to just change the behavior to respect the limit no matter what?

**Comment by gregkohn** [comment]
> Also interested in hearing how you settled on craft.viget.util as opposed to viget.util. (Fancy! I'm not opposed, just seems like a thing you must have had to choose)

@olets If i understand you correctly, extending the `craft` variable is how you get access to services. https://docs.craftcms.com/v3/extend/extending-twig.html#extend-the-global-craft-variable

File: tests/unit/FillInEntriesTest.php (hunk 5 above)
Comment (olets) [reply tests/unit/FillInEntriesTest.php:62]: sure and then only use `fillInEntries` if you know need more? makes sense! I think either way works

**Comment by olets** [comment]
Ah got it, didn't realize it's just the way it's done 👍

File: tests/fixtures/EntriesFixture.php (hunk 1 above)
Comment (davist11) [reply tests/fixtures/EntriesFixture.php]: Depends was enough, some sort of fixture magic

File: composer.json (hunk 3 above)
Comment (davist11) [reply composer.json]: Updated to `"vigetbasetests\\fixtures\\": "tests/fixtures/"` to match craft

File: tests/functional/PartialCest.php (hunk 2 above)
Comment (gregkohn) [reply tests/functional/PartialCest.php:20]: 🧠

**Review by gregkohn — APPROVED** [review]
👏

File: tests/fixtures/EntriesFixture.php (hunk 1 above)
Comment (olets) [reply tests/fixtures/EntriesFixture.php]: sweet!

**Pull request merged**